napari/_qt/qt_color_dialog.py

```diff
@@ -0,0 +1,257 @@
+from ..utils.colormaps.standardize_color import (
+    hex_to_name,
+    rgb_to_hex,
+    transform_color,
+)
+from .qt_modal import QtPopup
+from vispy.color import get_color_dict
+
+# matches any 3- or 4-tuple of int or float, with or without parens
+# captures the numbers into groups.
+# this is used to allow users to enter colors as e.g.: "(1, 0.7, 0)"
+rgba_regex = re.compile(
+    r"\(?([\d.]+),\s*([\d.]+),\s*([\d.]+),?\s*([\d.]+)?\)?"
+)
+
+TRANSPARENT = np.array([0, 0, 0, 0], np.float32)
+
+
+class QColorSwatchEdit(QWidget):
+    """A widget that combines a QColorSwatch with a QColorLineEdit.
+
+    emits a color_changed event with a 1x4 numpy array when the current color
+    changes.  Note, the "model" for the current color is the ``_color``
+    attribute on the QColorSwatch.
+    """
+
+    color_changed = Signal(np.ndarray)
+
+    def __init__(
+        self,
+        parent: Optional[QWidget] = None,
+        *,
+        initial_color: Optional[ColorType] = None,
+        tooltip: Optional[str] = None,
+    ):
+        """Create a new ColorSwatchEdit widget.
+
+        Parameters
+        ----------
+        parent : QWidget, optional
+            parent widget, by default None
+        initial_color : ColorType, optional
+            Starting color, by default None
+        tooltip : str, optional
+            Tooltip when hovering on the swatch,
+            by default 'click to set color'
+        """
+        super().__init__(parent=parent)
+        self.setObjectName('QColorSwatchEdit')
+
+        layout = QHBoxLayout()
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.setSpacing(6)
+        self.setLayout(layout)
+
+        self.lineEdit = QColorLineEdit(self)
+        self.lineEdit.editingFinished.connect(self._on_lineedit_edited)
+
+        self.colorSwatch = QColorSwatch(self, tooltip=tooltip)
+        self.colorSwatch.color_changed.connect(self._on_swatch_changed)
+        self.setColor = self.colorSwatch.setColor
+        if initial_color is not None:
+            self.setColor(initial_color)
+
+        layout.addWidget(self.colorSwatch)
+        layout.addWidget(self.lineEdit)
+
+    def color(self):
+        """Return the current color."""
+        return self.colorSwatch.color()
+
+    def _on_lineedit_edited(self):
+        """When the user hits enter or loses focus on the LineEdit widget."""
+        text = self.lineEdit.text()
+        rgb_match = rgba_regex.match(text)
+        if rgb_match:
+            text = [float(x) for x in rgb_match.groups() if x]
+        self.colorSwatch.setColor(text)
+
+    @Slot(np.ndarray)
+    def _on_swatch_changed(self, color: np.ndarray):
+        """Receive QColorSwatch change event, update the lineEdit, re-emit."""
+        self.lineEdit.setText(color)
+        self.color_changed.emit(color)
+
+
+class QColorSwatch(QFrame):
+    """A QFrame that displays a color.
+
+    The internal color representation (self._color) is always a 1x4 np.ndarray
+    """
```

Comment:
I think the convention, for better or worse, is to have the `__init__` parameters in this docstring, as well as the public attributes.

Reply:
really? I thought it was the other way around...  [pep 257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) seems to suggest that public methods and instance vars can go in the class docstring, but the constructor parameters go in the `__init__` docstring.  in any case, i can add the public stuff to the class docstring.

Reply:
That's disappointing that it goes against PEP 257, but it's what I've seen everywhere else, e.g. sklearn:

https://github.com/scikit-learn/scikit-learn/blob/244d118ef77a513b487b95b721179c395cbc1660/sklearn/ensemble/_forest.py#L850

Reply:
(And it makes sense for sphinx rendering...)

Reply:
yeah, that's a bummer... I don't love it actually.  but I guess the hoops that sphinx makes you jump through with `__init__` justifies it.